libc/string/string_errsig.c

```diff
@@ -175,3 +208,120 @@ TEST_GROUP_RUNNER(string_errsign)
 	RUN_TEST_CASE(string_errsign, strsignal_basic);
 	RUN_TEST_CASE(string_errsign, strsignal_real_time);
 }
+
+
+TEST_GROUP(string_perror);
+
+
+TEST_SETUP(string_perror)
+{
+	errno = 0;
+}
+
+
+TEST_TEAR_DOWN(string_perror)
+{
+}
+
+TEST(string_perror, perror_basic)
+{
+	char msg[] = "Some error message";
+	char *res;
+	char exp[100];
+	sprintf(exp, "Some error message: %s\n", strerror(0));
+
+	res = perrorToFile(msg);
+
+	TEST_ASSERT_EQUAL_STRING(exp, res);
+	free(res);
+}
+
+
+TEST(string_perror, perror_empty_message)
+{
+	TEST_IGNORE_MESSAGE("#929 issue");
+	char *res;
+	char exp[50];
+
+	errno = 31;
+	sprintf(exp, "%s\n", strerror(errno));
+	TEST_ASSERT_EQUAL_STRING(exp, res = perrorToFile(""));
+	free(res);
+}
+
+
+TEST(string_perror, perror_every_ascii)
+{
+	char *msg = testdata_createCharStr(257);
+	char expected[356];
+	char *res;
+
+	errno = 8;
+	res = perrorToFile(msg);
+
+	sprintf(expected, "%s: %s\n", msg, strerror(errno));
+	TEST_ASSERT_EQUAL_STRING(expected, res);
+	free(res);
+	free(msg);
+}
+
+
+TEST(string_perror, perror_huge_argument)
+{
+	const char *msg = testdata_hugeStr;
+	char *res;
+	char expected[testdata_hugeSize + 100];
+
+	errno = 42;
+	res = perrorToFile(msg);
+
+	sprintf(expected, "%s: %s\n", msg, strerror(errno));
+	TEST_ASSERT_EQUAL_STRING(expected, res);
+	free(res);
+}
+
+
+TEST(string_perror, perror_every_errno)
+{
+	char *old_msg;
+	char *new_msg;
+
+	errno = 0;
+
+	for (int i = 0; i < 150; i++) {
+		errno = i;
+		old_msg = perrorToFile("Some msg");
+		errno++;
+		new_msg = perrorToFile("Some msg");
+		errno--;
+		TEST_ASSERT_NOT_EQUAL(0, strcmp(old_msg, new_msg));
+		free(new_msg);
+		free(old_msg);
+	}
+}
+
+
+TEST(string_perror, perror_null_argument)
+{
+	TEST_IGNORE_MESSAGE("#929 issue");
+	char *msg;
+	msg = perrorToFile(NULL);
+	char exp[50];
+
+	strcpy(exp, strerror(errno));
+	strcat(exp, "\n");
+
+	TEST_ASSERT_EQUAL_STRING(exp, msg);
+	free(msg);
+}
+
+
+TEST_GROUP_RUNNER(string_perror)
+{
+	RUN_TEST_CASE(string_perror, perror_basic);
+	RUN_TEST_CASE(string_perror, perror_empty_message);
+	RUN_TEST_CASE(string_perror, perror_every_ascii);
+	RUN_TEST_CASE(string_perror, perror_huge_argument);
+	RUN_TEST_CASE(string_perror, perror_every_errno);
+	RUN_TEST_CASE(string_perror, perror_null_argument);
+}
```

Comment:
I'm considering adding more test cases, possibly for:

- extreme values of errno
- what happens when an invalid character is placed in msg
- a huge message (we could use memset)

Reply:
Huge message has already been checked in `perror_huge_argument`,
Invalid character was checked in `perror_every_ascii`,
and extreme values of errno were checked in `perror_every_errno`.
I might modify conditions if I misunderstood something.